docs/madr/decisions/011-rate-limit-policy.md

```diff
@@ -0,0 +1,294 @@
+allows limiting connections on the specific listener but without checking the origin of the request.
+ 
+## Considered Options
+ 
+* Create a MeshRateLimit with basic TCP rate limit support
+* Create a MeshHTTPRateLimit and MeshTCPRateLimit with basic TCP rate limit support
+* Create advanced TCP rate limiting per requested service
+
+## Decision Outcome
+ 
+Chosen option: create a MeshRateLimit with basic TCP rate limit support
+ 
+## Decision Drivers
+ 
+- Replace existing policies with new policy-matching compliant resources
+- Add support for TCP connection limiting at the basic level
+ 
+## Considered Options
+ 
+### Naming
+ 
+- **MeshRateLimit**
+- MeshHTTPRateLimit and MeshTCPRateLimit
+ 
+## Solution
+ 
+### Current configuration
+Below is a sample rate-limiting configuration.
+ 
+```yaml
+spec:
+ sources:
+   - match:
+       kuma.io/service: "redis_kuma-demo_svc_6379"   
+ destinations:
+   - match:
+       kuma.io/service: "demo-app_kuma-demo_svc_5000"
+ conf:
+   http:
+     requests: 5
+     interval: 10s
+     onRateLimit:
+       status: 423
+       headers:
+         - key: "x-Kuma-rate-limited"
+           value: "true"
+           append: true
+```
+We are using [Envoy's local rate limit](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter) which supports only HTTP configuration.
+ 
+The configuration translates to the Envoy configuration on the route and the listener filter:
+ 
+```yaml
+{
+ "match": {
+   "prefix": "/",
+   "headers": [
+   {
+     "name": "x-kuma-tags",
+     "safe_regex_match": {
+     "google_re2": {},
+     "regex": ".*&kuma.io/service=[^&]*redis_kuma-demo_svc_6379[,&].*"
+     }
+   }
+   ]
+ },
+ "route": {
+   "cluster": "localhost:5000",
+   "timeout": "0s"
+ },
+ "typed_per_filter_config": {
+   "envoy.filters.http.local_ratelimit": {
+   "@type": "type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit",
+   ...
+   }
+ }
+}
+ 
+```
+ 
+### Specification
+ 
+The differences between TCP and HTTP rate limiting in `targetRef` and `from` requires different validation of objects.
+Idea is to create one policy but with `oneof` and use different validation methods for them.
+ 
+#### **HTTP Rate Limit**
+ 
+Rate limiting can be configured on both HTTP connection managers and routes. Thanks to the header `x-kuma-tags`, which is propagated and set by Envoy, we can recognize the origin of the HTTP request.
+ 
+#### Top level
+ 
+Top-level targetRef can have all available kinds:
+```yaml
+targetRef:
+ kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+ name: ...
+```
+Rate limiting is an inbound policy so only `from` should be configured.
+ 
+#### From level
+ 
+```yaml
+from:
+ - targetRef:
+     kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+     name: ...
+```
+ 
+Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense (there is any route that a request originates from).
```

Comment:
```suggestion
Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense (there isn't a route that a request originates from).
```

Reply:
@michaelbeaumont for a parallel discussion out of scope of this MADR. How do we envision this in the future with route? This is a use case for route to be defined on the producer side (I want to define a rate-limit for a subset of my api).

Reply:
wouldn't you use MeshHTTPRoute in top-level targetRef to do so?